File: backend/app/models/spending.py
import uuid
from datetime import datetime
from sqlalchemy import String, Text, DateTime, ForeignKey, Numeric, Boolean, func
from sqlalchemy.orm import Mapped, mapped_column
from sqlalchemy.dialects.postgresql import UUID
from app.models.database import Base


class Spending(Base):
    __tablename__ = "spendings"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    conversation_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("conversations.id", ondelete="SET NULL"),
        nullable=True,
    )
    amount: Mapped[float] = mapped_column(Numeric(10, 2), nullable=False)
    category: Mapped[str] = mapped_column(String(20), nullable=False)
    note: Mapped[str | None] = mapped_column(Text, nullable=True)
    reaction: Mapped[str] = mapped_column(Text, nullable=False)
    chat_reaction: Mapped[str | None] = mapped_column(Text, nullable=True)
    chat_delivered: Mapped[bool] = mapped_column(Boolean, default=False)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now()
    )

File: app/lib/models/spending.dart
// app/lib/models/spending.dart
class Spending {
  final String id;
  final String? conversationId;
  final double amount;
  final String category;
  final String? note;
  final String reaction;
  final String? chatReaction;
  final bool chatDelivered;
  final DateTime createdAt;

  const Spending({
    required this.id, this.conversationId, required this.amount,
    required this.category, this.note, required this.reaction,
    this.chatReaction, required this.chatDelivered, required this.createdAt,
  });

  factory Spending.fromJson(Map<String, dynamic> json) => Spending(
    id: json['id'], conversationId: json['conversation_id'],
    amount: (json['amount'] as num).toDouble(), category: json['category'],
    note: json['note'], reaction: json['reaction'],
    chatReaction: json['chat_reaction'], chatDelivered: json['chat_delivered'] ?? false,
    createdAt: DateTime.parse(json['created_at']),
  );
}
